pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,188 @@
+			// PodLifeCycleEvent is ignored by the existing Generic PLEG as well.
+			// https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/pleg/generic.go#L88 and
+			// https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/pleg/generic.go#L273
+			klog.V(4).InfoS("Recieved Container Created Event", "CRI container event", event)
+		case runtimeapi.ContainerEventType_CONTAINER_STARTED_EVENT:
+			e.updatePodStatus(event)
+			e.eventChannel <- &PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerStarted, Data: event.ContainerId}
+			klog.V(4).InfoS("Recieved Container Started Event", "CRI container event", event)
+		case runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT:
+			e.updatePodStatus(event)
+			// In case the pod is deleted it is safe to generate both ContainerDied and ContainerRemoved events, just like in the case of
+			// Generic PLEG. https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/pleg/generic.go#L169
+			e.eventChannel <- &PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerDied, Data: event.ContainerId}
+			e.eventChannel <- &PodLifecycleEvent{ID: types.UID(event.PodSandboxMetadata.Uid), Type: ContainerRemoved, Data: event.ContainerId}
+			klog.V(4).InfoS("Recieved Container Deleted Event", "CRI container event", event)
+		}
+	}
+}
+
+func (e *EventedPLEG) updatePodStatus(event *runtimeapi.ContainerEventResponse) {
+
+	podID := types.UID(event.PodSandboxMetadata.Uid)
+	podName := event.PodSandboxMetadata.Name
+	podNamespace := event.PodSandboxMetadata.Namespace
+
+	timestamp := e.clock.Now()
+	status, err := e.runtime.GetPodStatus(podID, podName, podNamespace)
+
+	if err != nil {
+		// The existing generic PLEG is not disabled, instead the relisting is done less frequently. So,
+		// in case of error we rely on that generic PLEG relisting to correctly update the pod status.
+		if klog.V(6).Enabled() {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+	} else {
+		if klogV := klog.V(6); klogV.Enabled() {
+			klogV.InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.V(4).InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+		// Preserve the pod IP across cache updates if the new IP is empty.
+		// When a pod is torn down, kubelet may race with PLEG and retrieve
+		// a pod status after network teardown, but the kubernetes API expects
+		// the completed pod's IP to be available after the pod is dead.
+		status.IPs = e.getPodIPs(podID, status)
+	}
+
+	if event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT {
+		for _, sandbox := range status.SandboxStatuses {
+			if sandbox.Id == event.ContainerId {
```

Comment:
should never happen on containerd.. a sandbox might be a vm type... and containerd does not overload containerid and podsandbox ids.. a podsandbox has it's own id and it's own lifecycle in containerd.

Reply:
We are just updating kubelet's internal cache. This is very similar to how the [existing](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/pleg/generic.go#L396) Generic PLEG handles the pod deletion from the kubelet cache. In case of Generic PLEG, they will delete the pod from the cache if it doesn't show up during the relisting. While in our case,  we get explicit delete event from the runtime so we also go ahead and delete the pod from kube cache. We just make sure that the delete event we got was for the sandbox container and not the constituent container(s) of the pod before deleting the entire pod from the cache.

Reply:
ok.. my concern I guess is small windows / disconnects between the idea of the sandbox container being deleted and the sandbox being deleted..  from (https://github.com/containerd/containerd/blob/main/pkg/cri/server/sandbox_remove.go#L96-L112)
```
	// Delete sandbox container.
	if err := sandbox.Container.Delete(ctx, containerd.WithSnapshotCleanup); err != nil {
		if !errdefs.IsNotFound(err) {
			return nil, fmt.Errorf("failed to delete sandbox container %q: %w", id, err)
		}
		log.G(ctx).Tracef("Remove called for sandbox container %q that does not exist", id)
	}

	// Remove sandbox from sandbox store. Note that once the sandbox is successfully
	// deleted:
	// 1) ListPodSandbox will not include this sandbox.
	// 2) PodSandboxStatus and StopPodSandbox will return error.
	// 3) On-going operations which have held the reference will not be affected.
	c.sandboxStore.Delete(id)

	// Release the sandbox name reserved for the sandbox.
	c.sandboxNameIndex.ReleaseByKey(id)
```

Reply:
Din't quite get the concern that might be caused due to this. In the current implementation, the pod cache will be updated once the container with ID as that of sandbox ID, which seems natural (although we do understand that it actually doesn't indicate the deletion of the pod sandbox). The pod sandbox may still remain active though the sandbox container exits, but since our implementation is event-driven and we're not using relisting, this condition was used to update the cache.

Reply:
> Din't quite get the concern that might be caused due to this. In the current implementation, the pod cache will be updated once the container with ID as that of sandbox ID gets deleted

container runtime is containerd...
```
mike@ubnt:~/crictl-test$ sudo crictl run container-config.json sandbox-config.json 
301bd43e2bfb139188af9d5156c6be76fd7aace714423d2aa7f12a07d9dd66b2
mike@ubnt:~/crictl-test$ sudo crictl pods
POD ID              CREATED             STATE               NAME                NAMESPACE           ATTEMPT             RUNTIME
dcc2efa57c400       17 seconds ago      Ready               busybox-sandbox     default             1                   (default)
```
note pod id is dcc

```
mike@ubnt:~/crictl-test$ sudo crictl ps --all
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID
301bd43e2bfb1       busybox:1.35.0      21 seconds ago      Running             busybox             0                   dcc2efa57c400
```
note container id for container inside the pod is 301 and it's pod is dcc

> , which seems natural (although we do understand that it actually doesn't indicate the deletion of the pod sandbox). 

doesn't seem natural to me, it seems early, for example if the sequence for pod construction is create id, then network namespace then attach that to the pause container,.... during removal the sequenc will be remove the pause container then the network namespace.. then destroy meta store and return pod id in a transaction..

> The pod sandbox may still remain active though the sandbox container exits, but since our implementation is event-driven and we're not using relisting, this condition was used to just update the cache.

It's not a "may" thing the sandbox container is just one possible part of the sandbox, and is not a requirement for pod sandboxes. In fact there will come a time where there will be no sandbox container for sandboxed pods. Only a matter of time, we (mostly maksym) are currently working on a sandboxed feature in containerd.

There should be pod events to indicate the pod has exited. Unless we are saying we want kubelet to manage tracking all the checkpointed/inited parts of the pod sandbox, individually?

from the ctr (containerd tool) here you can see the pause container and the container:

```
mike@ubnt:~/crictl-test$ sudo ctr -n k8s.io c ls
CONTAINER                                                           IMAGE                               RUNTIME                  
301bd43e2bfb139188af9d5156c6be76fd7aace714423d2aa7f12a07d9dd66b2    docker.io/library/busybox:1.35.0    io.containerd.runc.v2    
dcc2efa57c400dbf98935f11e37ffb8c5544a3ae28bf530ef8a9fc5dc26e60d4    k8s.gcr.io/pause:3.7                io.containerd.runc.v2 
```
now kill the pause task, ls the containers showing the container is still there, and delete the pause container using the ctr tool:
```
mike@ubnt:~/crictl-test$ sudo ctr -n k8s.io t kill -s KILL dcc2efa57c400dbf98935f11e37ffb8c5544a3ae28bf530ef8a9fc5dc26e60d4
mike@ubnt:~/crictl-test$ sudo ctr -n k8s.io c ls
CONTAINER                                                           IMAGE                               RUNTIME                  
301bd43e2bfb139188af9d5156c6be76fd7aace714423d2aa7f12a07d9dd66b2    docker.io/library/busybox:1.35.0    io.containerd.runc.v2    
dcc2efa57c400dbf98935f11e37ffb8c5544a3ae28bf530ef8a9fc5dc26e60d4    k8s.gcr.io/pause:3.7                io.containerd.runc.v2    
mike@ubnt:~/crictl-test$ sudo ctr -n k8s.io c del dcc2efa57c400dbf98935f11e37ffb8c5544a3ae28bf530ef8a9fc5dc26e60d4
mike@ubnt:~/crictl-test$ sudo ctr -n k8s.io c ls
CONTAINER                                                           IMAGE                               RUNTIME                  
301bd43e2bfb139188af9d5156c6be76fd7aace714423d2aa7f12a07d9dd66b2    docker.io/library/busybox:1.35.0    io.containerd.runc.v2    
```
note pause container is deleted

now come in through CRI to see what's still around:
```
mike@ubnt:~/crictl-test$ sudo crictl ps -a
CONTAINER           IMAGE               CREATED             STATE               NAME                ATTEMPT             POD ID
301bd43e2bfb1       busybox:1.35.0      19 minutes ago      Exited              busybox             0                   dcc2efa57c400
mike@ubnt:~/crictl-test$ sudo crictl pods
```
note container has exited

check pod list
```
POD ID              CREATED             STATE               NAME                NAMESPACE           ATTEMPT             RUNTIME
dcc2efa57c400       19 minutes ago      NotReady            busybox-sandbox     default             1                   (default)
```
note pod is still there, and now in notready state


I suppose one could use the information about the pause exit as a point in time "update" to a cache moving the pod to the "exiting" state, if kubelet is going to accept said new state?

Reply:
Thanks @mikebrow for that perspective.